### PR TITLE
Refactored `src/` and cleaned up tests

### LINKED ISSUE #21: Refactoring `src`
Now that `src/DynamicNLPModels.jl` is quite long, we want to refactor the source code. I suggest

- `src/` contains `DynamicNLPModels.jl` and `utils.jl`, which are used accorss nonlinear and LQ models.
- `src/lq/` contains everything related to LQ problems
- `src/nonlinear/` contains everything related to nonlinear problems (it should be empty now. you can put an empty code for now and say it's TODO)

For restructuring the code, take a look at other packages, like `QuadraticModels.jl`

### PR DESCRIPTION
**Pull request by dlcole3**
The source code in `src/` is over 2000 lines long and needed to be refactored. I added two subdirectories, `nonlinear` and `lq`. `nonlinear` contains an empty Julia file that just has a TODO comment. `lq` now contains four files: `lq.jl`, `sparse.jl`, `dense.jl`, and `tools.jl`. `lq.jl` contains most of the data structures for the linear quadratic MPC problem. `sparse.jl` contains all the code for building the sparse models. `dense.jl` contains the code for building the dense and the dense-implicit models. `tools.jl` contains various extended functions and API functions, such as the `get_u` or `get_Q` functions, as well as the `mul!` extensions and `add_jtsj` functions.

I also updated the test folder so that testing the `mul!` extension and `add_jtsj!` function are done within `@testset` and which calls both `Arrays` and `CuArrays` the same.

### CODE REVIEW

File: src/lq/lq.jl [hunk 1]
```diff
@@ -0,0 +1,279 @@
+
```
Comment (frapac) [inline src/lq/lq.jl]: Suggestion: rename this file as `LinearQuadratic.jl`

File: test/runtests.jl [hunk 2]
```diff
@@ -124,6 +108,47 @@ function dynamic_data_to_CUDA(dnlp::LQDynamicData)
     )
 end
 
+function runtests(model, dnlp, lq_sparse, lq_dense, lq_sparse_from_data, lq_dense_from_data, N, ns, nu)
```
Comment (frapac) [inline test/runtests.jl:111]: Good idea!

File: src/nonlinear/nonlinear.jl [hunk 3]
```diff
@@ -0,0 +1 @@
+#TODO: Add support for general nonlinear problems
```
Comment (frapac) [inline src/nonlinear/nonlinear.jl]: I would remove this file and wait for the implementation of the nonlinear extension

File: src/lq/tools.jl [hunk 4]
```diff
@@ -0,0 +1,723 @@
...
+Query the solution `u` from the solver. If `K = nothing`, the solution for `u` is queried from `solution_ref.solution`
+
+If `K <: AbstractMatrix`, `solution_ref.solution` returns `v`, and `get_u` solves for `u` using the `K` matrix (and the `A` and `B` matrices if `lqdm <: DenseLQDynamicModel`)
+"""
+function get_u(
+solver_status,
+lqdm::SparseLQDynamicModel{T, V, M1, M2, M3, MK}
+) where {T, V <: AbstractVector{T}, M1 <: AbstractMatrix{T}, M2 <: AbstractMatrix{T}, M3 <: AbstractMatrix{T}, MK <: AbstractMatrix{T}}
+
+solution = solver_status.solution
```
Comment (frapac) [inline src/lq/tools.jl]: Identation!

**Review by frapac — CHANGES_REQUESTED** [review]
This refactoring is more than welcome. 
As a suggestion, I would suggest renaming the `lq` folder as `LinearQuadratic` (explicit is better than implicit)

Also: we should be careful to not break the indentation!

**Review by sshin23 — APPROVED** [review]
Looks good @dlcole3. The main source code and test scripts are much easier to read now

File: src/LinearQuadratic/dense.jl [hunk 5]
```diff
@@ -0,0 +1,1005 @@
...
+function DenseLQDynamicModel(dnlp::LQDynamicData{T,V,M}; implicit = false) where {T, V <: AbstractVector{T}, M  <: AbstractMatrix{T}, MK <: Union{Nothing, AbstractMatrix{T}}}
+if implicit
+    _build_implicit_dense_lq_dynamic_model(dnlp)
+else
+    _build_dense_lq_dynamic_model(dnlp)
+end
+end
+
+function DenseLQDynamicModel(
+s0::V,
```
Comment (frapac) [inline src/LinearQuadratic/dense.jl]: Indentation

File: src/LinearQuadratic/tools.jl [hunk 6]
```diff
@@ -0,0 +1,724 @@
...
+Query the solution `u` from the solver. If `K = nothing`, the solution for `u` is queried from `solution_ref.solution`
+
+If `K <: AbstractMatrix`, `solution_ref.solution` returns `v`, and `get_u` solves for `u` using the `K` matrix (and the `A` and `B` matrices if `lqdm <: DenseLQDynamicModel`)
+"""
+function get_u(
+    solver_status,
+    lqdm::SparseLQDynamicModel{T, V, M1, M2, M3, MK}
+) where {T, V <: AbstractVector{T}, M1 <: AbstractMatrix{T}, M2 <: AbstractMatrix{T}, M3 <: AbstractMatrix{T}, MK <: AbstractMatrix{T}}
+
+solution = solver_status.solution
```
Comment (frapac) [inline src/LinearQuadratic/tools.jl]: indentation

File: src/LinearQuadratic/tools.jl [hunk 7]
```diff
@@ -0,0 +1,724 @@
...
+
+return u
+end
+
+function get_u(
+    solver_status,
+    lqdm::DenseLQDynamicModel{T, V, M1, M2, M3, M4, MK}
+) where {T, V <: AbstractVector{T}, M1 <: AbstractMatrix{T}, M2 <: AbstractMatrix{T}, M3 <: AbstractMatrix{T}, M4 <: AbstractMatrix{T}, MK <: AbstractMatrix{T}}
+
+dnlp = lqdm.dynamic_data
```
Comment (frapac) [inline src/LinearQuadratic/tools.jl]: Indentation

File: src/LinearQuadratic/tools.jl [hunk 8]
```diff
@@ -0,0 +1,724 @@
...
+LinearAlgebra.axpy!(1, As0, s)
+
+Ks = _init_similar(dnlp.s0, size(K, 1), T)
+u = copy(v)
+for i in 1:N
+    LinearAlgebra.mul!(Ks, K, s[(1 + ns * (i - 1)):ns * i])
+    u[(1 + nu * (i - 1)):nu * i] .+= Ks
+end
+
+return u
```
Comment (frapac) [inline src/LinearQuadratic/tools.jl]: Indentation

File: src/LinearQuadratic/tools.jl [hunk 9]
```diff
@@ -0,0 +1,724 @@
...
+end
+
+function get_u(
+    solver_status,
+    lqdm::SparseLQDynamicModel{T, V, M1, M2, M3, MK}
+) where {T, V <: AbstractVector{T}, M1 <: AbstractMatrix{T}, M2 <: AbstractMatrix{T}, M3 <: AbstractMatrix{T}, MK <: Nothing}
+
+solution = solver_status.solution
```
Comment (frapac) [inline src/LinearQuadratic/tools.jl]: Indentation

File: src/LinearQuadratic/tools.jl [hunk 10]
```diff
@@ -0,0 +1,724 @@
...
+u = solution[(ns * (N + 1) + 1):end]
+return u
+end
+
+function get_u(
+    solver_status,
+    lqdm::DenseLQDynamicModel{T, V, M1, M2, M3, M4, MK}
+) where {T, V <: AbstractVector{T}, M1 <: AbstractMatrix{T}, M2 <: AbstractMatrix{T}, M3 <: AbstractMatrix{T}, M4 <: AbstractMatrix{T}, MK <: Nothing}
+
+return copy(solver_status.solution)
```
Comment (frapac) [inline src/LinearQuadratic/tools.jl]: Indentation

File: src/LinearQuadratic/tools.jl [hunk 11]
```diff
@@ -0,0 +1,724 @@
...
+Query the solution `s` from the solver. If `lqdm <: SparseLQDynamicModel`, the solution is queried directly from `solution_ref.solution`
+If `lqdm <: DenseLQDynamicModel`, then `solution_ref.solution` returns `u` (if `K = nothing`) or `v` (if `K <: AbstactMatrix`), and `s` is found form
+transforming `u` or `v` into `s` using `A`, `B`, and `K` matrices.
+"""
+function get_s(
+    solver_status,
+    lqdm::SparseLQDynamicModel{T, V, M1, M2, M3, MK}
+) where {T, V <: AbstractVector{T}, M1 <: AbstractMatrix{T}, M2 <: AbstractMatrix{T}, M3 <: AbstractMatrix{T}, MK <: Union{Nothing, AbstractMatrix}}
+
+solution = solver_status.solution
```
Comment (frapac) [inline src/LinearQuadratic/tools.jl]: Indentation

File: src/LinearQuadratic/tools.jl [hunk 12]
```diff
@@ -0,0 +1,724 @@
...
+s = solution[1:(ns * (N + 1))]
+return s
+end
+
+function get_s(
+    solver_status,
+    lqdm::DenseLQDynamicModel{T,V, M1, M2, M3, M4, MK}
+) where {T, V <: AbstractVector{T}, M1 <: AbstractMatrix{T}, M2 <: AbstractMatrix{T}, M3 <: AbstractMatrix{T}, M4 <: AbstractMatrix{T}, MK <: Union{Nothing, AbstractMatrix}}
+
+dnlp = lqdm.dynamic_data
```
Comment (frapac) [inline src/LinearQuadratic/tools.jl]: Indentation

File: src/LinearQuadratic/tools.jl [hunk 13]
```diff
@@ -0,0 +1,724 @@
...
+
+LinearAlgebra.mul!(As0, block_A, dnlp.s0)
+LinearAlgebra.axpy!(1, As0, s)
+
+return s
+end
+
+for field in fieldnames(LQDynamicData)
+method = Symbol("get_", field)
+@eval begin
```
Comment (frapac) [inline src/LinearQuadratic/tools.jl]: Indentation

File: src/LinearQuadratic/tools.jl [hunk 14]
```diff
@@ -0,0 +1,724 @@
...
+    @doc """
+        $($method)(LQDynamicData, row, col, val)
+        $($method)(SparseLQDynamicModel, row, col, val)
+        $($method)(DenseLQDynamicModel, row, col, val)
+    Set the value of entry $($(QuoteNode(field)))[row, col] to val for LQDynamicData, SparseLQDynamicModel.dynamic_data, or DenseLQDynamicModel.dynamic_data
+    """
+    $method(dyn_data::LQDynamicData, row, col, val) = (dyn_data.$field[row, col] = val)
+end
+@eval $method(dyn_model::SparseLQDynamicModel, row, col, val) = (dyn_model.dynamic_data.$field[row, col] = val)
```
Comment (frapac) [inline src/LinearQuadratic/tools.jl]: Indentation

File: src/LinearQuadratic/tools.jl [hunk 15]
```diff
@@ -0,0 +1,724 @@
...
+end
+
+function NLPModels.jac_structure!(
+    qp::DenseLQDynamicModel{T, V, M1, M2, M3},
+    rows::AbstractVector{<:Integer},
+    cols::AbstractVector{<:Integer},
+  ) where {T, V, M1<: Matrix, M2 <: Matrix, M3 <: Matrix}
+    count = 1
+    for j = 1:(qp.meta.nvar)
+      for i = 1:(qp.meta.ncon)
```
Comment (frapac) [inline src/LinearQuadratic/tools.jl]: Indentation: 4 spaces instead of 2 spaces

File: src/LinearQuadratic/tools.jl [hunk 16]
```diff
@@ -0,0 +1,724 @@
...
+
+function NLPModels.hess_coord!(
+    qp::DenseLQDynamicModel{T, V, M1, M2, M3},
+    x::AbstractVector{T},
+    vals::AbstractVector{T};
+    obj_weight::Real = one(eltype(x)),
+  ) where {T, V, M1 <: Matrix, M2 <: Matrix, M3 <: Matrix}
+    NLPModels.increment!(qp, :neval_hess)
+    count = 1
+    for j = 1:(qp.meta.nvar)
```
Comment (frapac) [inline src/LinearQuadratic/tools.jl:280]: Indentation: 4 spaces instead of 2 spaces

File: src/LinearQuadratic/tools.jl [hunk 17]
```diff
@@ -0,0 +1,724 @@
...
+end
+
+function NLPModels.jac_coord!(
+    qp::DenseLQDynamicModel{T, V, M1, M2, M3},
+    x::AbstractVector,
+    vals::AbstractVector,
+    ) where {T, V, M1 <: Matrix, M2 <: Matrix, M3 <: Matrix}
+    NLPModels.increment!(qp, :neval_jac)
+    count = 1
+    for j = 1:(qp.meta.nvar)
```
Comment (frapac) [inline src/LinearQuadratic/tools.jl:347]: Indentation: 4 spaces instead of 2 spaces

**Review by frapac — CHANGES_REQUESTED** [review]
There are still a few issues with the indentation. This is weird: have you used a formatter?

**Review by frapac — APPROVED** [review]
Great, I think the PR is ready to merge now!

**Pull request merged**